File: src/aoc24/day1/haskell/solution.hs
module Day1.Haskell.Solution
  ( runDay1
  )
where

import Data.List (sort)
import Data.Map (Map)
import Data.Map qualified as Map
import System.IO (readFile)
import Control.Exception (try, SomeException)

readInputFile :: IO [(Int, Int)]
readInputFile = do
  result <- try (readFile "inputs/day1/input.txt") :: IO (Either SomeException String)
  case result of
    Left e -> do
      putStrLn $ "Error reading file: " ++ show e
      return []
    Right contents -> do
      return $ map parseLine $ lines contents

parseLine :: String -> (Int, Int)
parseLine line =
  let tokens = words line
  in case tokens of
    [n1, n2] -> (read n1, read n2)
    _ -> error $ "Invalid line format: " ++ line

part1 :: [(Int, Int)] -> Int
part1 pairs =
  sum $
    zipWith
      (\n1 n2 -> abs (n1 - n2))
      (sort $ map fst pairs)
      (sort $ map snd pairs)

frequencies :: Ord a => [a] -> Map a Int
frequencies = foldr (\a -> Map.insertWith (+) a 1) Map.empty

part2 :: [(Int, Int)] -> Int
part2 pairs =
  let list1 = map fst pairs
      list2 = map snd pairs
      freq1 = frequencies list2
   in sum $ map (\n1 -> n1 * Map.findWithDefault 0 n1 freq1) list1

runDay1 :: IO ()
runDay1 = do
  putStrLn "Starting Day 1..."
  input <- readInputFile
  if null input 
    then putStrLn "No input data!"
    else do
      let p1 = part1 input
      let p2 = part2 input
      putStrLn $ "Part 1: " ++ show p1
      putStrLn $ "Part 2: " ++ show p2

File: src/aoc24/day1/solution.py
def part1():
    list1 = []
    list2 = []
    with open("inputs/day1/input.txt") as f:
        for _ in range(1000):
            line = f.readline()
            tokens = line.split("   ")
            list1.append(int(tokens[0]))
            list2.append(int(tokens[1]))

    list1 = sorted(list1)
    list2 = sorted(list2)
    diff = 0
    for i in range(1000):
        diff += abs(list1[i] - list2[i])
    return diff


def part2():
    list1 = []
    frequencies = {}
    with open("inputs/day1/input.txt") as f:
        for _ in range(1000):
            line = f.readline()
            tokens = line.split("   ")
            list1.append(int(tokens[0]))

            t2 = int(tokens[1])
            if t2 not in frequencies:
                frequencies[t2] = 0
            frequencies[t2] += 1
    sim_score = 0
    for i in list1:
        sim_score += i * frequencies.get(i, 0)

    return sim_score


if __name__ == "__main__":
    p1 = part1()
    print(f"Part 1: {p1}")

    p2 = part2()
    print(f"Part 2: {p2}")
